src/System.Memory/src/System/Buffers/ReadOnlySequence_helpers.cs

```diff
@@ -14,131 +14,114 @@ namespace System.Buffers
+                {
+                    Debug.Assert(positionObject is string);
 
-                data = (ReadOnlyMemory<T>)(object)(Unsafe.As<string>(startObject)).AsMemory();
-            }
-            else
-            {
-                data = default;
-                return false;
+                    data = (ReadOnlyMemory<T>)(object)(Unsafe.As<string>(positionObject)).AsMemory();
+                }
+                else // if (type == SequenceType.MemoryManager)
+                {
+                    Debug.Assert(positionObject is MemoryManager<T>);
+
+                    data = (Unsafe.As<MemoryManager<T>>(positionObject)).Memory;
+                }
             }
 
-            data = data.Slice(startIndex, length);
+            data = data.Slice(positionIndex, endIndex - positionIndex);
             return true;
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal ReadOnlyMemory<T> GetFirstBuffer(in SequencePosition start, in SequencePosition end)
+        private ReadOnlyMemory<T> GetFirstBuffer(in SequencePosition start, in SequencePosition end)
         {
-            SequenceType type;
-            int startIndex = 0;
-            int endIndex = 0;
+            GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out SequenceType type, out int startIndex, out int endIndex);
             object startObject = start.GetObject();
-            if (startObject != null)
-            {
-                GetTypeAndIndices(start.GetInteger(), end.GetInteger(), out type, out startIndex, out endIndex);
-            }
-            else
-            {
-                type = SequenceType.Empty;
-            }
-
-            int length = endIndex - startIndex;
-            object endObject = end.GetObject();
-
-            if (type != SequenceType.MultiSegment && type != SequenceType.Empty && startObject != endObject)
-                ThrowHelper.ThrowInvalidOperationException_EndPositionNotReached();
-
             ReadOnlyMemory<T> memory;
             if (type == SequenceType.MultiSegment)
             {
                 Debug.Assert(startObject is ReadOnlySequenceSegment<T>);
                 ReadOnlySequenceSegment<T> startSegment = Unsafe.As<ReadOnlySequenceSegment<T>>(startObject);
+
                 memory = startSegment.Memory;
-                if (startObject != endObject)
-                {
-                    length = memory.Length - startIndex;
-                }
+                if (startObject != end.GetObject())
+                    endIndex = memory.Length;
             }
             else if (type == SequenceType.Array)
             {
                 Debug.Assert(startObject is T[]);
 
                 memory = new ReadOnlyMemory<T>(Unsafe.As<T[]>(startObject));
             }
-            else if (type == SequenceType.MemoryManager)
+            else if (typeof(T) == typeof(char) && type == SequenceType.String)
```

Comment:
Why reorder?

Reply:
Because this section can be optimized by compiler if T is not char.
The last section is optimized by code (comparition is commented).
So now 1 or 2 section can be optimized.
If makes this section last then one section will be optimized always.
Now the last and some ty